### PR TITLE
KEYCLOAK-5162: Adding indices to speed up event queries

### PR DESCRIPTION
**Pull request by mariapavlova**

### CODE REVIEW

**Comment by mariapavlova** [comment]
We would like to have the indices created concurrently, otherwise updates are taking a long time to complete. Please advise

**Comment by stianst** [comment]
What do you mean by "indices created concurrently"?

**Comment by mariapavlova** [comment]
I meant something like this "CREATE INDEX CONCURRENTLY idx_event_entity_type ON event_entity (type);" as described in https://issues.jboss.org/browse/KEYCLOAK-5162

File: model/jpa/src/main/resources/META-INF/jpa-changelog-3.3.0.xml [hunk 1]
```diff
@@ -23,4 +23,19 @@
         </addColumn>
     </changeSet>
 
+    <changeSet author="maria.pavlova@quest.com" id="3.3.0">
```
Comment (hmlnarik) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-3.3.0.xml]: Id has to be unique, set it to e.g. "3.3.0-index-event"

**Review by hmlnarik — CHANGES_REQUESTED** [review]
I don't think we should have index for each of the columns. This would be counterproductive, the writes into this - potentially heavy loaded - table would be too expensive. It certainly makes sense to have events indexed by time. What are your use cases, what do you usually filter by?

**Comment by mariapavlova** [comment]
Thank you, can you please take a look at the ticket mentioned above? the use cases described there in details.

**Comment by stianst** [comment]
3.3 has now been released so this has to be added to 3.4 changelog. The ID also has to be unique as Hynek mentions.

@hmlnarik we support searching events based on all columns this PR adds indexes for so the indexes make sense to me. Can you comment?

**Comment by hmlnarik** [comment]
@stianst I see. In that case we should limit the max_results as we do now. 

Looking at the code, filtering mandates having realm ID. The indices hence have to be composite ones, having realm ID as the first component. Otherwise we'll be getting into similar performance troubles with multiple realms.

My other concern is regarding startup time should this PR be merged. For fresh installation, there is no issue. In existing installations, building the indices can easily take tens of minutes which will timeout Keycloak startup (see comments on the JIRA). The only way I can see this PR merged would be:

1. Either the PR would contain database-specific commands for all supported databases protected by preconditions (CREATE INDEX CONCURRENTLY for Postgres, CREATE INDEX xyz ONLINE for Oracle, CREATE INDEX xy LOCK NONE for mysql, ALTER INDEX xyz REBUILD WITH (ONLINE = ON) for myssql),
2. Or have a migration guide stating that the if there is already an existing database, the DBA has to apply the changeset manually, otherwise timeouts can occur.

@stianst WDYT?

**Comment by stianst** [comment]
I really don't like the idea of requiring manual steps to upgrade the DB. So option 1 is the only way to go IMO.

**Comment by hmlnarik** [comment]
Related PR: #6324

**Pull request closed without merging**